src/codec.rs

```diff
@@ -0,0 +1,29 @@
+//! This module contains traits to have a unified API across codecs.
+//!
+//! There are two traits defined, [`Codec`] and [`Links`]. Those are separate traits as the `Links`
+//! trait is not generic over a certain type.
+
+use cid::Cid;
+
+/// Each IPLD codec implementation should implement this Codec trait. This way codecs can be more
+/// easily exchanged or combined.
+pub trait Codec<T> {
+    /// The multicodec code of the IPLD codec.
+    const CODE: u64;
+    /// The error that is returned if encoding or decoding fails.
+    type Error;
+
+    /// Decode a slice into the desired type.
+    fn decode(bytes: &[u8]) -> Result<T, Self::Error>;
+    /// Encode a type into bytes.
+    fn encode(obj: &T) -> Result<Vec<u8>, Self::Error>;
+}
+
+/// Trait for returning the links of a serialized IPLD data.
+pub trait Links {
+    /// The error that is returned if the link extraction fails.
+    type Error;
+
+    /// Return all links (CIDs) that the given encoded data contains.
+    fn links(bytes: &[u8]) -> Result<Vec<Cid>, Self::Error>;
```

Comment:
This should return an iterator. Possibly just `Result<impl Iterator<Item = Cid>, Self::Error>` unless we now have a way to name anonymous types as associated types (IIRC, we don't?).

Reply:
`ipld-extract-links` returns a vector as it's not easy to return an iterator (I guess you could use channels or something similar). Hence I kept it simple and didn't want to make it seem like it's an iterator, although it's backed by a vector.

Reply:
We can just call `vec.into_iter()` on the returned vector. I'm suggesting that we return an iterator here because, e.g., it's pretty easy to implement streaming link extractors for most (all?) codecs.

Reply:
(even if we don't do that by default)